From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Jurgens <danielj@nvidia.com>
Date: Sun, 8 Feb 2026 22:14:07 +0000
Subject: [PATCH 168/254] virt: tegra_hv: move IVC layout declarations to
 tegra_hv_ivc.h

Move struct tegra_hv_guest_area, tegra_hv_notify_info,
tegra_hv_ivc_layout and the related function declarations
(tegra_hv_ivc_layout_init, tegra_hv_ivc_layout_release,
tegra_hv_get_max_qid, tegra_hv_add_ivc_interrupts,
tegra_hv_ivc_reserve_id, tegra_hv_ivc_unreserve_id) from
tegra_hv.h to tegra_hv_ivc.h.

This allows out-of-tree modules such as tegra_hv_nvscicom to
include tegra_hv_ivc.h instead of tegra_hv.h, avoiding
redefinition errors caused by overlapping declarations
between tegra_hv.h and hv-ivc.h.

tegra_hv.h now includes tegra_hv_ivc.h for backward
compatibility.

Bug 5628239

Change-Id: I8717661aa037de5d321bce870f407a410a4c036a
Signed-off-by: Dan Jurgens <danielj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3545240
Reviewed-by: svcacv <svcacv@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Aparna Das <aparnad@nvidia.com>
(cherry picked from commit 6875009f39ed9fc696a990b3377d2def495912f9)

diff --git a/include/soc/tegra/virt/tegra_hv.h b/include/soc/tegra/virt/tegra_hv.h
index 035d2546d28d7..6fefb333bf095 100644
--- a/include/soc/tegra/virt/tegra_hv.h
+++ b/include/soc/tegra/virt/tegra_hv.h
@@ -27,44 +27,6 @@
 /** @brief The maximum number of mempools supported by the PCT. */
 #define PCT_MAX_NUM_MEMPOOLS	120U
 
-/** @brief structure representing guest IVC area info */
-struct tegra_hv_guest_area {
-	/** @brief IO remapped shared memory address */
-	uintptr_t shmem;
-	/** @brief length of shared memory */
-	size_t length;
-};
-
-/** @brief structure representing IVC notification info */
-struct tegra_hv_notify_info {
-	/** @brief trap region base virtual address */
-	uintptr_t trap_region_base_va;
-	/** @brief trap region base IPA */
-	uint64_t trap_region_base_ipa;
-	/** @brief trap region end IPA */
-	uint64_t trap_region_end_ipa;
-	/** @brief trap region size */
-	uint64_t trap_region_size;
-	/** @brief MSI region base virtual address */
-	uintptr_t msi_region_base_va;
-	/** @brief MSI region base IPA */
-	uint64_t msi_region_base_ipa;
-	/** @brief MSI region end IPA */
-	uint64_t msi_region_end_ipa;
-	/** @brief MSI region size */
-	uint64_t msi_region_size;
-};
-
-/** @brief structure representing IVC layout */
-struct tegra_hv_ivc_layout {
-	/** @brief pointer to IVC info page */
-	const struct ivc_info_page *info;
-	/** @brief pointer to guest IVC area info array */
-	struct tegra_hv_guest_area *guest_ivc_info;
-	/** @brief notification info */
-	struct tegra_hv_notify_info notify;
-};
-
 /** @brief structure representing mempool cookie in hypervisor driver */
 struct tegra_hv_ivm_cookie {
 	/** @brief mempool base ipa address */
@@ -212,168 +174,6 @@ const struct ivc_info_page *tegra_hv_get_ivc_info(void);
  */
 int tegra_hv_get_vmid(void);
 
-/**
- * @brief          Reserve an IVC queue by ID
- * @param[in]      id Queue ID to reserve
- *
- * @retval         true If the queue was successfully reserved
- * @retval         false If already reserved or if the queue doesn't exist
- *
- * @pre
- *                 - Tegra hypervisor driver should have been initialized.
- *                 - This API should be invoked on virtual/hypervisor environment only.
- *
- * @post
- *                 - Queue will be marked reserved with reserved_by_external flag set
- *
- * @usage
- *                 - Used by tegra_hv_nvscicom to prevent double reservation of queues.
- *                 - Allowed context for the API call
- *                   - Interrupt handler: No
- *                   - Signal handler: N/A
- *                   - Thread-safe: Yes
- *                   - Async/Sync: Sync
- *                   - Re-entrant: No
- *                 - API Group
- *                   - Init: No
- *                   - Runtime: Yes
- *                   - De-Init: No
- */
-bool tegra_hv_ivc_reserve_id(uint32_t id);
-
-/**
- * @brief          Unreserve an IVC queue by ID
- * @param[in]      id Queue ID to unreserve
- *
- * @pre
- *                 - Tegra hypervisor driver should have been initialized.
- *                 - This API should be invoked on virtual/hypervisor environment only.
- *                 - Queue should have been reserved via tegra_hv_ivc_reserve_id.
- *
- * @post
- *                 - Queue reservation will be cleared if reserved_by_external flag was set
- *
- * @usage
- *                 - Used by tegra_hv_nvscicom to release the reservation.
- *                 - Allowed context for the API call
- *                   - Interrupt handler: No
- *                   - Signal handler: N/A
- *                   - Thread-safe: Yes
- *                   - Async/Sync: Sync
- *                   - Re-entrant: No
- *                 - API Group
- *                   - Init: No
- *                   - Runtime: Yes
- *                   - De-Init: No
- */
-void tegra_hv_ivc_unreserve_id(uint32_t id);
-
-/**
- * @brief          Initialize IVC layout structure
- * @param[in,out]  layout Pointer to layout structure to initialize
- * @param[in]      read_ivc_info Function pointer to read IVC info page address
- *
- * @retval         0 On success, negative error code on failure
- *
- * @pre
- *                 - layout must be a valid pointer
- *                 - read_ivc_info must be a valid function pointer
- *
- * @post
- *                 - layout structure will be initialized with IVC info
- *
- * @usage
- *                 - Allowed context for the API call
- *                   - Interrupt handler: No
- *                   - Signal handler: N/A
- *                   - Thread-safe: No
- *                   - Async/Sync: Sync
- *                   - Re-entrant: No
- *                 - API Group
- *                   - Init: Yes
- *                   - Runtime: No
- *                   - De-Init: No
- */
-int tegra_hv_ivc_layout_init(struct tegra_hv_ivc_layout *layout,
-			     int (*read_ivc_info)(uint64_t *ivc_info_page_pa));
-
-/**
- * @brief          Release IVC layout resources
- * @param[in]      layout Pointer to layout structure to release
- *
- * @pre
- *                 - layout should have been initialized via tegra_hv_ivc_layout_init
- *
- * @post
- *                 - All resources in layout will be released
- *
- * @usage
- *                 - Allowed context for the API call
- *                   - Interrupt handler: No
- *                   - Signal handler: N/A
- *                   - Thread-safe: No
- *                   - Async/Sync: Sync
- *                   - Re-entrant: No
- *                 - API Group
- *                   - Init: No
- *                   - Runtime: No
- *                   - De-Init: Yes
- */
-void tegra_hv_ivc_layout_release(struct tegra_hv_ivc_layout *layout);
-
-/**
- * @brief          Get maximum queue ID from IVC info
- * @param[in]      info Pointer to IVC info page
- *
- * @retval         Maximum queue ID found in IVC info page
- *
- * @pre
- *                 - info must be a valid pointer to IVC info page
- *
- * @usage
- *                 - Allowed context for the API call
- *                   - Interrupt handler: Yes
- *                   - Signal handler: N/A
- *                   - Thread-safe: Yes
- *                   - Async/Sync: Sync
- *                   - Re-entrant: Yes
- *                 - API Group
- *                   - Init: Yes
- *                   - Runtime: Yes
- *                   - De-Init: No
- */
-uint32_t tegra_hv_get_max_qid(const struct ivc_info_page *info);
-
-/**
- * @brief          Add interrupts property to device node
- * @param[in]      dev Device node to add property to
- * @param[in]      info Pointer to IVC info page
- * @param[in]      prop Property structure to use (caller must keep alive)
- *
- * @retval         0 On success, negative error code on failure
- *
- * @pre
- *                 - dev, info, prop must be valid pointers
- *
- * @post
- *                 - Interrupts property will be added to device node
- *
- * @usage
- *                 - Allowed context for the API call
- *                   - Interrupt handler: No
- *                   - Signal handler: N/A
- *                   - Thread-safe: No
- *                   - Async/Sync: Sync
- *                   - Re-entrant: No
- *                 - API Group
- *                   - Init: Yes
- *                   - Runtime: No
- *                   - De-Init: No
- */
-int tegra_hv_add_ivc_interrupts(struct device_node *dev,
-				const struct ivc_info_page *info,
-				struct property *prop);
-
 /** @} */
 
 #endif /* TEGRA_HV_H */
diff --git a/include/soc/tegra/virt/tegra_hv_ivc.h b/include/soc/tegra/virt/tegra_hv_ivc.h
index d88650786019f..27a1e3c9e9881 100644
--- a/include/soc/tegra/virt/tegra_hv_ivc.h
+++ b/include/soc/tegra/virt/tegra_hv_ivc.h
@@ -7,13 +7,215 @@
 #define TEGRA_HV_IVC_H
 
 #include <linux/of.h>
+#include <linux/types.h>
 #include <linux/version.h>
+#include <soc/tegra/virt/syscalls.h>
 
 /**
  * @defgroup hypervisor_ivc_framework Hypervisor IVC Framework
  * @{
  */
 
+/** @brief structure representing guest IVC area info */
+struct tegra_hv_guest_area {
+	/** @brief IO remapped shared memory address */
+	uintptr_t shmem;
+	/** @brief length of shared memory */
+	size_t length;
+};
+
+/** @brief structure representing IVC notification info */
+struct tegra_hv_notify_info {
+	/** @brief trap region base virtual address */
+	uintptr_t trap_region_base_va;
+	/** @brief trap region base IPA */
+	uint64_t trap_region_base_ipa;
+	/** @brief trap region end IPA */
+	uint64_t trap_region_end_ipa;
+	/** @brief trap region size */
+	uint64_t trap_region_size;
+	/** @brief MSI region base virtual address */
+	uintptr_t msi_region_base_va;
+	/** @brief MSI region base IPA */
+	uint64_t msi_region_base_ipa;
+	/** @brief MSI region end IPA */
+	uint64_t msi_region_end_ipa;
+	/** @brief MSI region size */
+	uint64_t msi_region_size;
+};
+
+/** @brief structure representing IVC layout */
+struct tegra_hv_ivc_layout {
+	/** @brief pointer to IVC info page */
+	const struct ivc_info_page *info;
+	/** @brief pointer to guest IVC area info array */
+	struct tegra_hv_guest_area *guest_ivc_info;
+	/** @brief notification info */
+	struct tegra_hv_notify_info notify;
+};
+
+/**
+ * @brief          Initialize IVC layout structure
+ * @param[in,out]  layout Pointer to layout structure to initialize
+ * @param[in]      read_ivc_info Function pointer to read IVC info page address
+ *
+ * @retval         0 On success, negative error code on failure
+ *
+ * @pre
+ *                 - layout must be a valid pointer
+ *                 - read_ivc_info must be a valid function pointer
+ *
+ * @post
+ *                 - layout structure will be initialized with IVC info
+ *
+ * @usage
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: No
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: Yes
+ *                   - Runtime: No
+ *                   - De-Init: No
+ */
+int tegra_hv_ivc_layout_init(struct tegra_hv_ivc_layout *layout,
+			     int (*read_ivc_info)(uint64_t *ivc_info_page_pa));
+
+/**
+ * @brief          Release IVC layout resources
+ * @param[in]      layout Pointer to layout structure to release
+ *
+ * @pre
+ *                 - layout should have been initialized via tegra_hv_ivc_layout_init
+ *
+ * @post
+ *                 - All resources in layout will be released
+ *
+ * @usage
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: No
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: No
+ *                   - De-Init: Yes
+ */
+void tegra_hv_ivc_layout_release(struct tegra_hv_ivc_layout *layout);
+
+/**
+ * @brief          Get maximum queue ID from IVC info
+ * @param[in]      info Pointer to IVC info page
+ *
+ * @retval         Maximum queue ID found in IVC info page
+ *
+ * @pre
+ *                 - info must be a valid pointer to IVC info page
+ *
+ * @usage
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: Yes
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: Yes
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: Yes
+ *                 - API Group
+ *                   - Init: Yes
+ *                   - Runtime: Yes
+ *                   - De-Init: No
+ */
+uint32_t tegra_hv_get_max_qid(const struct ivc_info_page *info);
+
+/**
+ * @brief          Add interrupts property to device node
+ * @param[in]      dev Device node to add property to
+ * @param[in]      info Pointer to IVC info page
+ * @param[in]      prop Property structure to use (caller must keep alive)
+ *
+ * @retval         0 On success, negative error code on failure
+ *
+ * @pre
+ *                 - dev, info, prop must be valid pointers
+ *
+ * @post
+ *                 - Interrupts property will be added to device node
+ *
+ * @usage
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: No
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: No
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: Yes
+ *                   - Runtime: No
+ *                   - De-Init: No
+ */
+int tegra_hv_add_ivc_interrupts(struct device_node *dev,
+				const struct ivc_info_page *info,
+				struct property *prop);
+
+/**
+ * @brief          Reserve an IVC queue by ID
+ * @param[in]      id Queue ID to reserve
+ *
+ * @retval         true If the queue was successfully reserved
+ * @retval         false If already reserved or if the queue doesn't exist
+ *
+ * @pre
+ *                 - Tegra hypervisor driver should have been initialized.
+ *                 - This API should be invoked on virtual/hypervisor environment only.
+ *
+ * @post
+ *                 - Queue will be marked reserved with reserved_by_external flag set
+ *
+ * @usage
+ *                 - Used by tegra_hv_nvscicom to prevent double reservation of queues.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: No
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: Yes
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
+ */
+bool tegra_hv_ivc_reserve_id(uint32_t id);
+
+/**
+ * @brief          Unreserve an IVC queue by ID
+ * @param[in]      id Queue ID to unreserve
+ *
+ * @pre
+ *                 - Tegra hypervisor driver should have been initialized.
+ *                 - This API should be invoked on virtual/hypervisor environment only.
+ *                 - Queue should have been reserved via tegra_hv_ivc_reserve_id.
+ *
+ * @post
+ *                 - Queue reservation will be cleared if reserved_by_external flag was set
+ *
+ * @usage
+ *                 - Used by tegra_hv_nvscicom to release the reservation.
+ *                 - Allowed context for the API call
+ *                   - Interrupt handler: No
+ *                   - Signal handler: N/A
+ *                   - Thread-safe: Yes
+ *                   - Async/Sync: Sync
+ *                   - Re-entrant: No
+ *                 - API Group
+ *                   - Init: No
+ *                   - Runtime: Yes
+ *                   - De-Init: No
+ */
+void tegra_hv_ivc_unreserve_id(uint32_t id);
+
 /** @brief structure representing ivc queue cookie in hypervisor driver */
 struct tegra_hv_ivc_cookie {
 	/** @brief irq linked to the ivc queue */
-- 
2.43.0

